File: bin/MadGraph5_aMCatNLO/XtoAAto4G/XtoAAto4G_gridpack/config/SCRAM/findDependencies.pl
#!/usr/bin/env perl
# findDependencies.pl
#

use strict;
use File::Find ();
use Getopt::Long;
Getopt::Long::config('bundling_override');

my %options;
GetOptions(\%options,'rel=s', 'arch=s','scramroot=s');

my $rel=$options{'rel'};
my $scramarch=$options{'arch'} || $ENV{SCRAM_ARCH};
my $scramroot=$options{'scramroot'} || $ENV{SCRAMV1_ROOT};

if ($scramroot eq "")
{
  $scramroot=`sh -v scram arch 2>&1 |  grep 'SCRAMV1_ROOT=' | sed 's|;.*||;s|SCRAMV1_ROOT=||;s|"||g' | sed -e "s|'||g"`;
  chomp $scramroot;
}

# Set the variable $File::Find::dont_use_nlink if you're using AFS,
# since AFS cheats.

# for the convenience of &wanted calls, including -eval statements:
use vars qw/*name *dir *prune/;
*name   = *File::Find::name;
*dir    = *File::Find::dir;
*prune  = *File::Find::prune;

sub wanted;
sub doexec ($@);


use Cwd ();
my $cwd = Cwd::cwd();

my %uses;
my %usedby;
my @prod2src;
my $beginning="";

# Traverse desired filesystems
my $directory="${rel}/tmp";
File::Find::find({wanted => \&wanted}, $directory);

open F1,">${rel}/etc/dependencies/uses.out";
foreach my $key (keys %uses) {
    print F1 "$key $uses{$key}\n";
}
close F1;

open F1,">${rel}/etc/dependencies/usedby.out";
foreach my $key (keys %usedby) {
    print F1 "$key $usedby{$key}\n";
}
close F1;

open F1,">${rel}/etc/dependencies/prod2src.out";
foreach my $dep (@prod2src) {print F1 "$dep";}
close F1;

&pythonDeps($rel);
&buildFileDeps($rel, $scramarch, $scramroot);

exit;

sub wanted {
  if ($name=~/^.*(\.dep)\z/os){doexec(0, 'cat','{}');}
  elsif ($name=~/\/scram-prod2src[.]txt$/os){
    open F1, $name;
    push @prod2src, <F1>;
    close F1;
  }
}

sub doexec ($@) {
    my $ok = shift;
    my @command = @_; # copy so we don't try to s/// aliases to constants

    my $getnext=0;
    my $depname;
    open F1, $name;
    while (my $l=<F1>) {
    	chomp($l);
	last if ($l=~/^[^:]+ :\s*$/o);
	$l=~s/\s*\\$//o;
	my @sp1=split(' ',$l);
	next if ( scalar(@sp1) == 0);
	next if ( length($sp1[0]) < 4);
	my @sp2=split('/',$sp1[0]);
	my $tsp1="";
	my $foundsrc=0;
	foreach my $t (@sp2) {
	    $tsp1.="$t/" if ( $foundsrc==1);
	    $foundsrc=1 if ( $t eq "src" ); 
	}
	chop($tsp1);
	next if ( $tsp1 eq "");
	if ( $getnext==1 ) {
	    $depname=$tsp1;
	    $getnext=0;
	}
	else{
	    if ( $sp1[0] =~ /^tmp\//o ) {
		if ( $sp1[0] =~/(\.o|\.cc)\s*:$/o) {
		    $getnext=1;
		}
	    }
	    else{
		if ( $sp1[0] =~ /^src/o ) {
		    $uses{$depname}.="$tsp1 ";
		    $usedby{$tsp1}.="$depname ";
		}
	    }
	}
    }
    close F1;
    chdir $cwd; #sigh
    chdir $File::Find::dir;
    return !$?;
}

sub pythonDeps()
{
  my $rel=shift;
  my %cache=();
  foreach my $fname (`find ${rel}/src -name "*.py" -type f`)
  {
    chomp $fname;
    my $file=$fname;
    $file=~s/^${rel}\/+src\/+//o;
    if($fname!~/\/python\//o){next;}
    foreach my $line (`grep 'import ' $fname`)
    {
      chomp $line;
      if ($line=~/^\s*#/){next;}
      if ($line=~/^\s*from\s+([^\s]+)\s+import\s+/)
      {
        foreach my $x (&import2CMSSWDir($1,\%cache))
        {
          $cache{usedby}{$x}{$file}=1;
  	  $cache{uses}{$file}{$x}=1;
        }
      }
      elsif ($line=~/^\s*import\s+([^\s]+)\s*/)
      {
        foreach my $x (&import2CMSSWDir($1,\%cache))
        {
          $cache{usedby}{$x}{$file}=1;
	  $cache{uses}{$file}{$x}=1;
        }
      }
    }
  }
  foreach my $type ("uses","usedby")
  {  
    my $ref;
    open $ref,">${rel}/etc/dependencies/py${type}.out";
    foreach my $x (sort keys %{$cache{$type}}){print $ref "$x ",join(" ",sort keys %{$cache{$type}{$x}}),"\n";}
    close($ref);
  }
}

sub import2CMSSWDir()
{
  my ($str,$cache)=@_;
  my @pyfiles=();
  foreach my $s (split ',',$str)
  {
    $s=~s/\./\//og;
    if (exists $cache->{pymodule}{$s}){push @pyfiles,$cache->{pymodule}{$s};}
    elsif(!exists $cache->{noncmsmodule}{$s})
    {
      if (-f "${rel}/python/${s}.py")
      {
	$s=~/^([^\/]+\/+[^\/]+)\/+(.+)$/o;
	$cache->{pymodule}{$s}="${1}/python/${2}.py";
	push @pyfiles,"${1}/python/${2}.py";
      }
      else{$cache->{noncmsmodule}{$s}=1;}
    }
  }
  return @pyfiles;
}

sub buildFileDeps()
{
  my $rel=shift;
  my $arch=shift;
  my $scramroot=shift;

  unshift @INC,"$scramroot", "${scramroot}/src";
  my $pcache={};
  eval ("use Cache::CacheUtilities;");
  if(!$@){$pcache=&Cache::CacheUtilities::read("${rel}/.SCRAM/${arch}/ProjectCache.db.gz");}
  else{die "Unable to find Cache/CacheUtilities.pm PERL module.";}  
  my %cache=();
  foreach my $dir (sort keys %{$pcache->{BUILDTREE}})
  {
    if ($pcache->{BUILDTREE}{$dir}{SUFFIX} ne ""){next;}
    if (scalar(@{$pcache->{BUILDTREE}{$dir}{METABF}}) == 0){next;}
    my $bf=$pcache->{BUILDTREE}{$dir}{METABF}[0]; $bf=~s/src\///;
    $cache{dirs}{$dir}=$bf;
    my $pack=$dir;
    my $class=$pcache->{BUILDTREE}{$dir}{CLASS};
    if ($class=~/^(LIBRARY|CLASSLIB)$/){$pack=$pcache->{BUILDTREE}{$dir}{PARENT};}
    $cache{packs}{$pack}=$dir;
  }
  foreach my $dir (keys %{$cache{dirs}}){&updateBFDeps($dir,$pcache,\%cache,"");}
  foreach my $type ("uses","usedby")
  {  
    my $ref;
    open $ref,">${rel}/etc/dependencies/bf${type}.out";
    foreach my $x (sort keys %{$cache{$type}}){print $ref "$x ",join(" ",sort keys %{$cache{$type}{$x}}),"\n";}
    close($ref);
  }
}

sub updateBFDeps()
{
  my ($dir,$pcache,$cache)=@_;
  my $bf=$cache->{dirs}{$dir};
  if (exists $cache->{uses}{$bf}){return;}
  $cache->{uses}{$bf}={};
  foreach my $pack (keys %{$pcache->{BUILDTREE}{$dir}{RAWDATA}{DEPENDENCIES}})
  {
    if (exists $cache->{packs}{$pack})
    {
      my $xdata=$cache->{packs}{$pack};
      &updateBFDeps($xdata,$pcache,$cache);      
      $xdata=$cache->{dirs}{$xdata};
      $cache->{uses}{$bf}{$xdata}=1;
      $cache->{usedby}{$xdata}{$bf}=1;
      foreach my $xdep (keys %{$cache->{uses}{$xdata}})
      {
	$cache->{uses}{$bf}{$xdep}=1;
        $cache->{usedby}{$xdep}{$bf}=1;    
      }
    }
  }
}

File: bin/MadGraph5_aMCatNLO/ewkwpjjjj_5f_LO/ewkwpjjjj_5f_LO_gridpack/config/SCRAM/findDependencies.py
#!/usr/bin/env python3

import sys, re, os, json, gzip
from argparse import ArgumentParser
sys.path.append(os.environ["SCRAM_TOOL_HOME"])
from SCRAM.Core.Utils import readProducts
from SCRAM.Configuration.ConfigArea import ConfigArea

parser = ArgumentParser()
parser.add_argument('-rel')
parser.add_argument('-arch')
parser.add_argument('-scramroot')
args = parser.parse_args()

rel = args.rel

global name
uses = {}
usedby = {}
prod2src = []
directory = rel + "/tmp"


def doexec():
  global uses
  getnext = 0
  depname = ""
  reCC = re.compile(r'(\.o|\.cc)\s*:$')
  re1 = re.compile(r'^[^:]+ :\s*$')
  re2 = re.compile(r'\s*\\$')
  with open(name, 'r', encoding="utf-8") as file:
    for l in file:
      l = l.rstrip('\n')
      if re1.search(l): break
      l = re2.sub(r'', l)
      sp1 = l.split()
      if len(sp1) == 0: continue
      if len(sp1[0]) < 4: continue
      sp2 = sp1[0].split('/')
      tsp1 = ""
      foundsrc = 0
      for t in sp2:
        if foundsrc == 1: tsp1 += "%s/" % t
        if t == "src": foundsrc = 1
      tsp1 = tsp1[:-1]
      if tsp1 == "": continue
      if getnext == 1:
        depname = tsp1
        getnext = 0
      else:
        if sp1[0][:4] == "tmp/":
          if reCC.search(sp1[0]):
            getnext = 1
        else:
          if sp1[0][:3] == "src":
            if depname not in uses:
              uses[depname] = "%s " % tsp1
            else:
              uses[depname] += "%s " % tsp1
            if tsp1 not in usedby:
              usedby[tsp1] = "%s " % depname
            else: 
              usedby[tsp1] += "%s " % depname


def write2File(path, data, type_= None, prod2src=None):
  with open(path, 'w') as file:
    if type_:
      if type_ not in data: data[type_] = {}
      for x in sorted(data[type_]):
        file.write("%s %s\n" % (x, " ".join(sorted(data[type_][x]))))
    elif prod2src:
      for dep in data:
        file.write(str(dep))
    else:
      for key, value in sorted(data.items()):
        file.write("%s %s\n" % (key, value))


def createCache(match, cache, file):
  for x in import2CMSSWDir(match.group(1), cache):
    if "usedby" not in cache: cache["usedby"] = {}
    if "uses" not in cache: cache["uses"] = {}
    if x not in cache["usedby"]: cache["usedby"][x] = {}
    if file not in cache["uses"]: cache["uses"][file] = {}
    if file not in cache["usedby"][x]: cache["usedby"][x][file] = {}
    if x not in cache["uses"][file]: cache["uses"][file][x] = {}
    cache["usedby"][x][file] = 1
    cache["uses"][file][x] = 1


def pythonDeps(rel):
  cache = {}
  for root, dirs, files in os.walk("%s/src/" % rel):
    for filename in files:
      fpath = os.path.join(root, filename)
      if re.search(r'.py$', fpath):
        file = fpath
        file = re.sub(r"^%s\/+src\/+" % rel, r'', file)
        if not re.search(r'\/python\/', fpath):
          continue
        with open(fpath, 'r', encoding="utf-8") as f:
          for line in f.readlines():
            if 'import ' in line:
              line = line.rstrip('\n')
              if re.search(r'^\s*#', line):
                continue
              match_from_import = re.search(r'^\s*from\s+([^\s]+)\s+import\s+', line)
              match_import = re.search(r'^\s*import\s+([^\s]+)\s*', line)
              if match_from_import:
                createCache(match_from_import, cache, file)
              elif match_import:
                createCache(match_import, cache, file)
  for type_ in ("uses","usedby"):
    write2File("%s/etc/dependencies/py%s.out" % (rel, type_), cache, type_)

def import2CMSSWDir(str, cache):
  pyfiles = []
  if "pymodule" not in cache: cache["pymodule"] = {}
  if "noncmsmodule" not in cache: cache["noncmsmodule"] = {}
  for s in str.split(","):
    s = re.sub(r'\.', r'/', s)
    if s in cache["pymodule"]:
      pyfiles.append(cache["pymodule"][s])
    elif s not in cache["noncmsmodule"]:
      if os.path.exists("%s/python/%s.py" % (rel, s)):
        match = re.search(r'^([^\/]+\/+[^\/]+)\/+(.+)$', s)
        if match:
          cache["pymodule"][s] = "%s/python/%s.py" % (match.group(1), match.group(2))
          pyfiles.append("%s/python/%s.py" % (match.group(1), match.group(2)))
      else: cache["noncmsmodule"][s] = 1
  return pyfiles


def updateBFDeps(dir, pcache, cache):
  bf = cache["dirs"][dir]
  if bf in cache["uses"]: return
  cache["uses"][bf] = {}
  deps = {}
  if "USE" in pcache[dir]:
    for use in pcache[dir]["USE"]:
      deps[use] = 1
  if "BUILDPRODUCTS" in pcache[dir]:
    for type in pcache[dir]["BUILDPRODUCTS"]:
      for prod in pcache[dir]["BUILDPRODUCTS"][type]:
        if "USE" not in pcache[dir]["BUILDPRODUCTS"][type][prod]:
          continue
        for use in pcache[dir]["BUILDPRODUCTS"][type][prod]["USE"]:
          deps[use] = 1
  for pack in deps:
    if pack in cache["dirs"]:
      updateBFDeps(pack, pcache, cache)
      xbf = cache["dirs"][pack]
      cache["uses"][bf][xbf] = 1
      if xbf not in cache["usedby"]:
          cache["usedby"][xbf] = {}
      cache["usedby"][xbf][bf] = 1
      for xdep in cache["uses"][xbf]:
        cache["uses"][bf][xdep] = 1
        cache["usedby"][xdep][bf] = 1
  return


def buildFileDeps(rel):
  area = ConfigArea()
  area.location(rel)
  pcache = readProducts(area)
  cache = {"dirs": {}, "uses": {}, "usedby": {}}
  for dir in sorted(pcache.keys()):
    bf = os.path.join(dir, "BuildFile.xml")
    if dir not in cache["dirs"]: cache["dirs"][dir] = {}
    cache["dirs"][dir] = bf
  for dir in cache["dirs"]:
    updateBFDeps(dir, pcache, cache)
  del cache["dirs"]
  for type_ in ("uses", "usedby"):
    write2File("%s/etc/dependencies/bf%s.out" % (rel, type_), cache, type_)


for root, dirs, files in os.walk(directory):
  for filename in files:
    name = os.path.join(root, filename)
    if filename[-4:] == ".dep":
        doexec()
    elif filename == "scram-prod2src.txt":
      import fileinput
      for line in fileinput.input(name):
        prod2src.append(line)


write2File(rel + "/etc/dependencies/uses.out", uses)
write2File(rel + "/etc/dependencies/usedby.out", usedby)
write2File(rel + "/etc/dependencies/prod2src.out", prod2src, prod2src=True)
pythonDeps(rel)
buildFileDeps(rel)
